packages/material-ui-unstyled/src/BadgeUnstyled/BadgeUnstyled.d.ts

```diff
@@ -0,0 +1,152 @@
+        as: React.ElementType;
+        styleProps?: Omit<BadgeUnstyledTypeMap<P, D>['props'], 'components' | 'componentsProps'>;
+      };
+      badge?: {
+        as?: React.ElementType;
+        styleProps?: Omit<BadgeUnstyledTypeMap<P, D>['props'], 'components' | 'componentsProps'>;
+      };
+    };
+    /**
+     * Wrapped shape the badge should overlap.
+     * @default 'rectangular'
+     */
+    overlap?: 'rectangular' | 'circular';
+    /**
+     * The content rendered within the badge.
+     */
+    badgeContent?: React.ReactNode;
+    /**
+     * The badge will be added relative to this node.
+     */
+    children?: React.ReactNode;
+    /**
+     * Override or extend the styles applied to the component.
+     */
+    classes?: {
+      /** Styles applied to the root element. */
+      root?: string;
+      /** Styles applied to the badge `span` element. */
+      badge?: string;
+      /** Styles applied to the root element if `color="primary"`. */
+      colorPrimary?: string;
+      /** Styles applied to the root element if `color="secondary"`. */
+      colorSecondary?: string;
+      /** Styles applied to the root element if `color="error"`. */
+      colorError?: string;
+      /** Styles applied to the root element if `variant="dot"`. */
+      dot?: string;
+      /** Styles applied to the root element if `variant="standard"`. */
+      standard?: string;
+      /** Styles applied to the root element if `anchorOrigin={{ 'top', 'right' }} overlap="rectangular"`. */
+      anchorOriginTopRightRectangular?: string;
+      /** Styles applied to the root element if `anchorOrigin={{ 'bottom', 'right' }} overlap="rectangular"`. */
+      anchorOriginBottomRightRectangular?: string;
+      /** Styles applied to the root element if `anchorOrigin={{ 'top', 'left' }} overlap="rectangular"`. */
+      anchorOriginTopLeftRectangular?: string;
+      /** Styles applied to the root element if `anchorOrigin={{ 'bottom', 'left' }} overlap="rectangular"`. */
+      anchorOriginBottomLeftRectangular?: string;
+      /** Styles applied to the root element if `anchorOrigin={{ 'top', 'right' }} overlap="circular"`. */
+      anchorOriginTopRightCircular?: string;
+      /** Styles applied to the root element if `anchorOrigin={{ 'bottom', 'right' }} overlap="circular"`. */
+      anchorOriginBottomRightCircular?: string;
+      /** Styles applied to the root element if `anchorOrigin={{ 'top', 'left' }} overlap="circular"`. */
+      anchorOriginTopLeftCircular?: string;
+      /** Styles applied to the root element if `anchorOrigin={{ 'bottom', 'left' }} overlap="circular"`. */
+      anchorOriginBottomLeftCircular?: string;
+      /** Pseudo-class to the badge `span` element if `invisible={true}`. */
+      invisible?: string;
+    };
+    /**
+     * The color of the component. It supports those theme colors that make sense for this component.
+     * @default 'default'
+     */
+    color?: 'primary' | 'secondary' | 'default' | 'error';
```

Comment:
Considering these components are completely styling agnostic what purpose those this prop serve? Shouldn't this be implemented by the styled variant? Same concern with the `variant` prop or any prop that is tied to some notion of a "theme".

Reply:
That's a good point, I don't think the `Unstyled` components should have the styling props. @oliviertassinari @mbrookes do you agree too on this?

Reply:
I agree with Sebastian, it could be an issue for developers that want to add their own prop, and extend the types. I guess it would be interesting to explore what are the alternatives and to weight the pros & cons.

Reply:
I've extracted the `color` prop to the styled version of the component in https://github.com/mnajdova/material-ui/pull/18 I don't think we can extract the variant prop, as it is used inside the logic of the component, which in my opinion should indicate that it is not strictly styling prop. From all other props we can extract the `overlap` prop as well, as it is strictly styling prop. For the rest I don't think we can as they are used in the logic of the component. Let me know what are your thoughts

Reply:
Great, thanks for looking into it. The change the most interesting seems to be the public API: https://github.com/mnajdova/material-ui/pull/18/files#diff-26877d58ee4ebaa271700aae39d84f0ce250e8d01c677024bbf0d53f46159b1b. It's basically two new props: `styleProps` and `extendClasses` to generate new class names. Do we need the unstyled components to have this capability or should host the logic in the styled version?

Reply:
The only reason why I introduced these props in the unstyled component is default values for the props. We even have some transient prop, like invisible that depend on other props values, so we need to consider these when applying the changes to have these values of the props. I’ll draft different version tomorrow without doing changes on the unstyled component API so we can compare.

Reply:
Here is the alternative https://github.com/mnajdova/material-ui/pull/19 - this PR is just removing the color prop in the BadgeUnstyled and fixes some bugs there. 

I think this approach has many downsides, if you take a look at the changes in the `Badge` component, it has to know the internal logic of the `UnstyledBadge` in order for example to correctly apply some additional styles - defaults of the props, invisible calculations etc. I think this component is a good example that this will probably create problems down the road, as code will need to be duplicated between the components. I would vote if we can go with an approach similar to https://github.com/mnajdova/material-ui/pull/18 - we can of course polish the API, but I think that is much safer than the second approach. The biggest problem is that, if someone uses the unstyled component, they will have to do all these things by themself.

Reply:
I might be leaning toward mnajdova#19 over mnajdova#18 because the necessary duplication of logic seems specific to the Badge (isn't true for the other components) and because it seems that we didn't yet identify a use case for this new API on the unstyled components. No strong point of view, it would be great to get @eps1lon & @mbrookes perspectives.

Reply:
I mean it somewhat flows into why we create an unstyled version in the first place. I would prefer we focus on the emotion migration first (planned for v5) and then look at possible unstyled components (can be a normal feature release).

Reply:
This is a valid point, adding unstyled component can be a feature that we add anytime, shouldn't be a breaking change. And it may help us move faster, as the emotion migration is fairly scoped and can be done easily in most of the cases. However, for adding the unstyled components, it would mean that we will need a second iteration on all components, which can be slower, but at least we won't be forced to do it fast and block the v5 release because of it.

Reply:
> then look at possible unstyled components (can be a normal feature release).

Interesting point, why did we do unstyled components at the same time? I see a couple of factors:

- it's as important as the migration to emotion, if not more, so might want to start with unstyled first. At least, it's how it was presented in #20012
- there is a risk that doing emotion then unstyled might not work, regarding another major. By doing both at the same time we hedge it.
- doing the migration to emotion requires refactoring the component, this seems to be an opportunity to be more efficient by refactoring directly to the final state we aim for.

Now, I would like to also raise that we talked about leveraging this churn opportunity to bake TypeScript and CSS string templates in. However, we concluded that doing them at the same time would increase the risk that the whole change fail without gaining much efficiency, so the idea was put aside.

Reply:
> it's as important as the migration to emotion, if not more, so might want to start with unstyled first. At least, it's how it was presented in #20012

I can see that it is the first feature we have there, so let's not cut that part then.. Let's start with mnajdova#18 and I will prepare PR to do the same for the Slider as well, and we can start from there. Sounds good?

Reply:
> it's as important as the migration to emotion, if not more

We just cannot v5 without getting rid of our styling solution. Otherwise we're piling on technical debt. However, we can release v5 without unstyled. We can always release them in a SemVer minor release. We cannot change the styling engine in a SemVer minor.

Reply:
Also: It is concerning that we're unable to split this PR into two parts. This indicates an accidental coupling. However, the primary use case of the unstyled package is to be agnostic to the styling engine. So far it does not seem like it.

If we can't even decouple this PR at a reasonable speed then the story for library users is even worse.

Reply:
> If we can't even decouple this PR at a reasonable speed then the story for library users is even worse.

This is double work, it's not that it is not possible. We can move faster if we do this in one go. Plus, this is a good first use case of the unstyled components.

Reply:
> Sounds good?

It sounds good, I personally prefer we don't introduce a new API on the unstyled components until we learn more about the problem faced by developers when using them. But I also value that if we struggle to style these unstyled components, then, their design might be possible to improve.

Reply:
This has been resolved.

Reply:
> This is double work, it's not that it is not possible. 

But that's a bad thing, no? It's actually bewildering to me how this would require double the work compared to a single PR. That is deeply concerning.

I don't understand why we create this component if
1. we don't know what problem it solves ("until we learn more about the problem faced by developers when using them.")
1. It requires double the work

"Speed" isn't a scalar unit. It requires a direction. Just because the magnitude is big does not mean that it's a good thing. It seems to me we don't have any picture on what this unstyled component is supposed to do. At least considering all the information that was given to me.

Reply:
It's double work in the means that, for moving to emotion you have the steps to map the classes generated, move the styling props to styleProps. Then for creating the unstyled components, you have to split the classes between unstyled + styled, and move the logic we've added to the core component to the unstyled. Making these changes in one go is simpler.

The styled component should contain only the business logic of the component, without the styling logic, but it needs to map the props somehow to classes so that it is useful and can be used for applying styles on top of it. (The core components in the end are pretty simpler wrapper around these components) My point was that, when you do the changes in one go, it's simpler to reason about these things, than doing the changes once only in the core component, and then going over the changes again in order to create the unstyled version and pick which changes should go there (at least this is my opinion, and maybe I am wrong, but as so far I've been the one doing these changes, I cannot use any other feedback at this moment about whether this is true or not).

The main reason for unstyled components is that we can ship the functionality for the components, without using Material Design, and without any opinion as to which styling mechanism developers should use, which can help us in the growth of the user base.

Reply:
> we don't know what problem it solves ("until we learn more about the problem faced by developers when using them.")

@eps1lon This was a reference to the new APIs introduced in mnajdova#18: the `styleProps` and `extendClasses` props.

> It requires double the work

We were talking about the difference between these two approaches, right? 1. emotion and unstyled in **one** PR vs. 2. doing emotion and unstyled in **two** PRs.

Reply:
Again, I am ok with any option, as long as we can move faster at this moment. I got the impression that unstyled is even more important than emotion, so that's why I said it would be faster if we do them in one PR. We can split them, and do the unstyled separately if you think that will help us

Reply:
I personally think that we should bundle unstyled and emotion to move faster and to increase the chance of getting both right.

Reply:
How about we try both options in the next migrated component? At least, this will help us have a better idea of:

- The time spent doing the emotion migration only
- The time spent doing the unstyled migration only
- The time spent doing the emotion+unstyled migration

Reply:
I also wonder if we shouldn't handle the migration of the documentation independently.

Reply:
Let's try to push this PR as is, and I think that https://github.com/mui-org/material-ui/pull/23841 is a good candidate for only swapping the styling engine. We can then decide what will work better. @eps1lon would that work?